wagtail/wagtailadmin/static_src/wagtailadmin/js/page-editor.js

```diff
@@ -361,80 +361,67 @@ $(function() {
-                success: function(data, textStatus, request) {
-                    if (request.getResponseHeader('X-Wagtail-Preview') == 'ok') {
-                        if (enhanced) {
-                            var frame = previewDoc.getElementById('preview-frame');
-
-                            frame = frame.contentWindow || frame.contentDocument.document || frame.contentDocument;
-                            frame.document.open();
-                            frame.document.write(data);
-                            frame.document.close();
-
-                            var hideTimeout = setTimeout(function() {
-                                previewDoc.getElementById('loading-spinner-wrapper').className += ' remove';
-                                clearTimeout(hideTimeout);
-                            });
-
- // just enough to give effect without adding discernible slowness
-                        } else {
-                            previewDoc.open();
-                            previewDoc.write(data);
-                            previewDoc.close();
-                        }
-
-                    } else {
-                        previewWindow.close();
-                        disableDirtyFormCheck();
-                        document.open();
-                        document.write(data);
-                        document.close();
-                    }
-                },
-
-                error: function(xhr, textStatus, errorThrown) {
-                    /* If an error occurs, display it in the preview window so that
-                    we aren't just showing the spinner forever. We preserve the original
-                    error output rather than giving a 'friendly' error message so that
-                    developers can debug template errors. (On a production site, we'd
-                    typically be serving a friendly custom 500 page anyhow.) */
-
-                    previewDoc.open();
-                    previewDoc.write(xhr.responseText);
-                    previewDoc.close();
-                }
-            });
+    //
+    // Preview
+    //
+    // In order to make the preview truly reliable, the preview page needs
+    // to be perfectly independent from the edit page,
+    // from the browser perspective. To pass data from the edit page
+    // to the preview page, we send the form after each change
+    // and save it inside the user session.
+
+    var $previewButton = $('.action-preview'), $form = $('#page-edit-form');
+    var previewUrl = $previewButton.data('action');
+
+    function setPreviewData() {
+        return $.ajax({
+            url: previewUrl,
+            method: 'POST',
+            data: new FormData($form[0]),
+            processData: false,
+            contentType: false
+        });
+    }
 
-        }
+    if ($previewButton.data('auto-update')) {
+        var setPreviewDataTimeout = -1;
+
+        // Form data is changed when field values are changed (change event),
```

Comment:
I think we should avoid sending any preview requests until the user clicks the preview button. Is that possible?